hacks/xteevee.man
.TH XScreenSaver 1 "19-Nov-99" "X Version 11"
.SH NAME
xteevee - Simulate various TV problems
.SH SYNOPSIS
.B xteevee
[\-display \fIhost:display.screen\fP] [\-window] [\-root] [\-install]
[\-visual \fIvisual\fP] 
[\-cycle] [\-no-cycle]
[\-static] [\-no-static] [\-roll] [\-no-roll] [\-bars] [\-no-bars] 
.SH DESCRIPTION
.I xteevee
simulates various television problems, fondly re-creating the days when
computers were fun and I spent my time with an Atari 400 hooked up to a
black-and-white TV I found in someone's trash.
.PP
When run,
.I xteevee
picks a mode to display from the available options.  If modes are explicitly
set on the command-line, they override the implicit options and will be
chosen among exclusively.  If modes are explicitly excluded from the
command-line, the they are simply removed from contention.

The image that it manipulates will be grabbed from the portion of
the screen underlying the window, or from the system's video input,
or from a random file on disk, as indicated by
the \fIgrabDesktopImages\fP, \fIgrabVideoFrames\fP,
and \fIchooseRandomImages\fP options in the \fI~/.xscreensaver\fP
file; see
.BR xscreensaver-demo (1)
for more details.
.PP
See also
.BR xanalogtv (MANSUFFIX)
for a more sophisticated (but more graphics-intensive) implementation
of this concept.
.SH OPTIONS
.I xteevee
accepts the following options:
.TP 8
.B \-window
Draw on a newly-created window.  This is the default.
.TP 8
.B \-root
Draw on the root window.
.TP 8
.B \-install
Install a private colormap for the window.
.TP 8
.B \-visual \fIvisual\fP
Specify which visual to use.  Legal values are the name of a visual class,
or the id number (decimal or hex) of a specific visual.
.TP 8
.B \-cycle
Cycle through all the available modes.  This is the default.
.TP 8
.B \-no-cycle
Don't cycle modes.
.TP 8
.B \-static
Explicitly select the "static" mode.
.TP 8
.B \-no-static
Explicitly forbid the "static" mode.
.TP 8
.B \-roll
Explicitly select the "vertical roll" mode.
.TP 8
.B \-no-roll
Explicitly forbid the "vertical roll" mode.
.TP 8
.B \-bars
Explicitly select the "color bars" mode.
.TP 8
.B \-no-bars
Explicitly forbid the "color bars" mode.
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.SH X RESOURCES
Notable X resources supported include the following:
.PP
.BR static,
.BR roll
and
.BR bars
are booleans, defining which modes are to be implicitly available if
.I xteevee
is run without command-line mode selections.
.BR cycle
is a boolean, defining if
.I xteevee
should cycle through each available mode or pick one and stick with it.
.BR delayMode
and
.BR delayBetween
are integers, defining how long each mode should be displayed and how
long the time between them should be if
.BR cycle
is on.
.BR staticSignal
is an integer, defining the strength of the signal during the "static"
mode, as a percentage.
.SH BUGS
.I xteevee
should simulate more TV problems.
.SH SEE ALSO
.BR X (1),
.BR xanalogtv (MANSUFFIX),
.BR apple2 (MANSUFFIX),
.BR bsod (MANSUFFIX),
.BR xscreensaver (1),
.BR xscreensaver\-demo (1),
.BR xscreensaver\-getimage (1)
.SH COPYRIGHT
Copyright \(co 1999 by Greg Knauss.  Permission to use, copy, modify, 
distribute, and sell this software and its documentation for any purpose is 
hereby granted without fee, provided that the above copyright notice appear 
in all copies and that both that copyright notice and this permission notice
appear in supporting documentation.  No representations are made about the 
suitability of this software for any purpose.  It is provided "as is" without
express or implied warranty.  If you've gotten this far, then you're obviously
looking for something to do.  Go here: http://www.teevee.org.
.SH AUTHOR
Greg Knauss <greg@eod.com>, 19-Nov-1999.


po/ChangeLog
2002-11-25  Kostas Papadimas <pkst@gmx.net>

	* el.po: Added Greek translation

2002-11-22  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Updated Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-17  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Added Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-06  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-24  Christian Meyer  <chrisime@gnome.org>

	* de.po: Updated German translation.

2002-09-23  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-22  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-20  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-19  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-17  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-16  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-08  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.
			This should fix the "error" on the statuspages.

2002-09-07  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-05  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-03  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmål) translation.

2002-08-29  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-08-29  Tõivo Leedjärv  <toivo@linux.ee>

	* et.po: Fixed Estonian translation.

2002-08-29  Christian Neumair  <christian-neumair@web.de>

	* de.po: Updated German translation.

2002-08-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Partial update of Danish translation.

2002-08-23  Andras Timar  <timar@gnome.huy

	* hu.po: Added Hungarian translation (partial).

2002-08-13  Christian Neumair  <christian-neumair@web.de>

	* de.po: Started German translation.

2002-08-11  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.

2002-07-28  Christophe Merlet <christophe@merlet.net>

	* fr.po: Updated French translation from contribution of
	Sun G11n <gnome_int_l10n@ireland.sun.com>.

2002-07-22  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-07-12  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po,wa.po: Updated Vietnamese and Walloon files

2002-06-26  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po: Fixed a duplicated string.

2002-06-24  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-25  Christian Rose  <menthos@menthos.com>

	* fr.po, pl.po, ru.po, uk.po, wa.po: Added from xscreensaver module.

2002-06-21  Changwoo Ryu  <cwryu@debian.org>

	* ko.po: Added Korean translation from Young-Ho Cha
	<ganadist@chollian.net>.

2002-06-19  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-06-18  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO, xscreensaver.pot: Updated from 4.0.5 release.
	
2002-06-16  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO: Added extra info. I will take all translations from here, you
	don't need to send me it by mail.

2002-06-16  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Added Vietnamese file

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-12  Abel Cheung  <maddog@linux.org.hk>

	* zh_TW.po: New traditional Chinese translation.

2002-06-09  Takeshi Aihana <aihana@gnome.gr.jp>

	* ja.po: Added as Japanese translation.

2002-06-06  Christian Rose  <menthos@menthos.com>

	* sv.po: Added from xscreensaver module.

2002-06-06  Ole Laursen  <olau@hardworking.dk>

	* da.po: Added and halfway-updated from xscreensaver module.

2002-06-06  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

	* et.po: Added from xscreensaver module.

2002-06-06  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po,nl.po,no.po,pt.po,sk.po,zh_CN.po: Added from xscreensaver
	module.
	* xscreensaver.pot: Added from latest release (4.04.2).
	* INFO: Added info about this module.

2002-06-04  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Added Dutch translation.

2002-05-26  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-25  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Backported Simplified Chinese translation from HEAD.

2002-05-19  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.
						
2002-05-17  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Added Simplified Chinese translation by
	He Qiangqiang <carton@linux.net.cn>.

2002-05-11  Stanislav Visnovsky  <visnovsky@nenya.ms.mff.cuni.cz>

        * sk.po: Backported Slovak translation from HEAD.
	
2002-05-01  GermÃ¡n Poo-CaamaÃ±o  <gpoo@ubiobio.cl>

	* es.po: Updated Spanish translation from
	Guillermo Movia <gmovia@hotpop.com> and converted to UTF-8

2002-04-29  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* wa.po: Updated Walloon file

2002-04-27  Duarte Loreto <happyguy_pt@hotmail.com>

        * pt.po: Added Portuguese translation.

2002-04-26  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

        * et.po: Updated Estonian translation.

2002-04-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-04-21  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmÃ¥l) translation.


hacks/xteevee.c
/*****************************************************************************
 *                                                                           *
 * xteevee -- TV good... TV good...                                          *
 *                                                                           *
 * Copyright (c) 1999 Greg Knauss (greg@eod.com)                             *
 *                                                                           *
 * Permission to use, copy, modify, distribute, and sell this software and   *
 * its documentation for any purpose is hereby granted without fee, provided *
 * that the above copyright notice appear in all copies and that both that   *
 * copyright notice and this permission notice appear in supporting          *
 * documentation.  No representations are made about the suitability of this *
 * software for any purpose.  It is provided "as is" without express or      *
 * implied warranty.                                                         *
 *                                                                           *
 *****************************************************************************/


/* Changelog *****************************************************************

	1.0.0	19991119	Initial release

*/


/* Includes ******************************************************************/
#include "screenhack.h"
#include "colorbars.h"
#include <X11/Xutil.h>
#include <time.h>
#include <sys/time.h>


/* Defines *******************************************************************/
#define XTEEVEE_NAME                    "XTeeVee"
#define XTEEVEE_MODE_EXCLUDE            0
#define XTEEVEE_MODE_INCLUDE_IMPLICIT   1
#define XTEEVEE_MODE_INCLUDE_EXPLICIT   2
#define XTEEVEE_ARG_STATIC              "static"
#define XTEEVEE_ARG_STATIC_SIGNAL       "staticSignal"
#define XTEEVEE_ARG_ROLL                "roll"
#define XTEEVEE_ARG_BARS                "bars"
#define XTEEVEE_ARG_CYCLE               "cycle"
#define XTEEVEE_ARG_DELAY_MODE          "delayMode"
#define XTEEVEE_ARG_DELAY_BETWEEN       "delayBetween"
#define XTEEVEE_STATIC_COLOR_COUNT      6
#define XTEEVEE_STATIC_TILE_COUNT       16


/* Globals *******************************************************************/
char *progclass  = XTEEVEE_NAME;
char *defaults[] =
{
	"*" XTEEVEE_ARG_STATIC        ": true",
	"*" XTEEVEE_ARG_STATIC_SIGNAL ": 50",
	"*" XTEEVEE_ARG_ROLL          ": true",
	"*" XTEEVEE_ARG_BARS          ": true",
	"*" XTEEVEE_ARG_CYCLE         ": true",
	"*" XTEEVEE_ARG_DELAY_MODE    ": 30",
	"*" XTEEVEE_ARG_DELAY_BETWEEN ": 3",
	NULL
};
XrmOptionDescRec options[] =
{
 { "-"    XTEEVEE_ARG_STATIC,"._" XTEEVEE_ARG_STATIC,XrmoptionNoArg,"true" },
 { "-no-" XTEEVEE_ARG_STATIC,"."  XTEEVEE_ARG_STATIC,XrmoptionNoArg,"false" },
 { "-"    XTEEVEE_ARG_ROLL  ,"._" XTEEVEE_ARG_ROLL  ,XrmoptionNoArg,"true" },
 { "-no-" XTEEVEE_ARG_ROLL  ,"."  XTEEVEE_ARG_ROLL  ,XrmoptionNoArg,"false" },
 { "-"    XTEEVEE_ARG_BARS  ,"._" XTEEVEE_ARG_BARS  ,XrmoptionNoArg,"true" },
 { "-no-" XTEEVEE_ARG_BARS  ,"."  XTEEVEE_ARG_BARS  ,XrmoptionNoArg,"false" },
 { "-"    XTEEVEE_ARG_CYCLE ,"."  XTEEVEE_ARG_CYCLE ,XrmoptionNoArg,"true" },
 { "-no-" XTEEVEE_ARG_CYCLE ,"."  XTEEVEE_ARG_CYCLE ,XrmoptionNoArg,"false" },
 { NULL                     ,NULL                   ,0             ,NULL }
};


/* Functions *****************************************************************/

/* Get the forground pixel ================================================= */
void xteevee_FreeColorForeground(Display* x_Disp,XWindowAttributes* x_WinAttr,
      GC x_Gc)
{
	XGCValues x_GcVal;

	if (XGetGCValues(x_Disp,x_Gc,GCForeground,&x_GcVal) != 0)
	{
		XFreeColors(x_Disp,x_WinAttr->colormap,&x_GcVal.foreground,1,
		 0);
	}
}

/* Static ================================================================== */
void xteevee_Static(Display* x_Disp,Window x_Win,XWindowAttributes* x_WinAttr,
      time_t hack_Time,Pixmap hack_Pm)
{
	GC        x_Gc[XTEEVEE_STATIC_COLOR_COUNT];
        unsigned long pixels[XTEEVEE_STATIC_COLOR_COUNT];
        XImage   *xim = 0;
        char     *orig_bits = 0;
	XGCValues x_GcVal;
	int       signal_Strength;
	XColor    color_Color;
	int       color_Index;
	int       tile_Index;
	Pixmap    tile_Tile[XTEEVEE_STATIC_TILE_COUNT];
	int       tile_X;
	int       tile_Y;
	int       tile_Width;
	int       tile_Height;
	char      tile_Used[XTEEVEE_STATIC_TILE_COUNT/2+1];
	int       tile_Selected;

	/* Get any extra arguments */
	signal_Strength = get_integer_resource(XTEEVEE_ARG_STATIC_SIGNAL,
	 "Integer");

	/* Build the GCs */
	color_Color.flags = DoRed|DoGreen|DoBlue;
	for (color_Index = 0;color_Index < XTEEVEE_STATIC_COLOR_COUNT;
	 color_Index++)
	{
		color_Color.red = color_Color.green = color_Color.blue =
		 (((double)color_Index+1)/XTEEVEE_STATIC_COLOR_COUNT)*65535;
		if (!x_WinAttr->colormap ||
                    !XAllocColor(x_Disp,x_WinAttr->colormap,&color_Color))
		{
			/* NOTE: I have no idea what to do here.  Why would
			         this fail? */
		}
                pixels[color_Index] = color_Color.pixel;
		x_GcVal.foreground = color_Color.pixel;
		x_Gc[color_Index] = XCreateGC(x_Disp,x_Win,GCForeground,
		 &x_GcVal);
	}

	/* Build the tiles */
	for (tile_Index = 0;tile_Index < XTEEVEE_STATIC_TILE_COUNT;
	 tile_Index++)
	{
		if (signal_Strength == 0)
		{
			/* NOTE: Checking XQueryBestTile() returns tiles that
			         are too regular -- you can see patterns
			         emerge. */
			tile_Width = (random()%128)+64;
			tile_Height = (random()%128)+64;
		}
		else
		{
			tile_Width = x_WinAttr->width;
			tile_Height = x_WinAttr->height;
		}
		tile_Tile[tile_Index] = XCreatePixmap(x_Disp,x_Win,tile_Width,
		 tile_Height,x_WinAttr->depth);
		XCopyArea(x_Disp,hack_Pm,tile_Tile[tile_Index],x_Gc[0],0,0,
		 x_WinAttr->width,x_WinAttr->height,0,0);

		if (signal_Strength == 0)
                  {
                    if (!xim)
                      {
                        xim = XCreateImage (x_Disp, x_WinAttr->visual,
                                            x_WinAttr->depth,
                                            (x_WinAttr->depth == 1
                                             ? XYPixmap : ZPixmap),
                                            0, 0,
                                            x_WinAttr->width,
                                            x_WinAttr->height,
                                            8, 0);
                        
                        xim->data = (char *) malloc (xim->bytes_per_line *
                                                     xim->height);
                      }
                  }
                else if (xim)
                  {
                    memcpy (xim->data, orig_bits,
                            xim->bytes_per_line * xim->height);
                  }
                else
                  {
                    xim = XGetImage (x_Disp, tile_Tile[tile_Index], 0, 0,
                                     x_WinAttr->width, x_WinAttr->height, ~0L,
                                     (x_WinAttr->depth == 1
                                      ? XYPixmap : ZPixmap));
                    orig_bits = (char *) malloc (xim->bytes_per_line *
                                                 xim->height);
                    memcpy (orig_bits, xim->data,
                            xim->bytes_per_line * xim->height);
                  }

                for (tile_Y = tile_Height-1;tile_Y >= 0;tile_Y--)
                  for (tile_X = tile_Width-1;tile_X >= 0;tile_X--)
                    if (random()%100 > signal_Strength)
                      XPutPixel(xim,tile_X,tile_Y,
                                pixels[random()%XTEEVEE_STATIC_COLOR_COUNT]);
                XPutImage(x_Disp,tile_Tile[tile_Index],x_Gc[0],xim,
                          0,0,0,0,x_WinAttr->width,x_WinAttr->height);
	}

        if (xim) XDestroyImage (xim);
        if (orig_bits) free (orig_bits);

	/* Go! */
	memset(tile_Used,-1,sizeof(tile_Used));
	if (hack_Time > 0)
	{
		hack_Time += time(NULL);
	}
	while ((time(NULL) < hack_Time) || (hack_Time == 0))
	{
		/* Pick a tile */
		do
		{
			tile_Selected = random()%XTEEVEE_STATIC_TILE_COUNT;
			for (tile_Index = 0;tile_Index < sizeof(tile_Used);
			 tile_Index++)
			{
				if (tile_Used[tile_Index] == tile_Selected)
				{
					tile_Selected = -1;
					break;
				}
			}
		} while (tile_Selected == -1);
		memmove(tile_Used,tile_Used+1,sizeof(tile_Used)-1);
		tile_Used[sizeof(tile_Used)-1] = tile_Selected;

		/* Set it */
		XSetWindowBackgroundPixmap(x_Disp,x_Win,
		 tile_Tile[tile_Selected]);
		XClearWindow(x_Disp,x_Win);

		XSync(x_Disp,0);
		screenhack_handle_events(x_Disp);
		usleep(25000);
	}

	/* Free everything */
	for (color_Index = 0;color_Index < XTEEVEE_STATIC_COLOR_COUNT;
	 color_Index++)
	{
		xteevee_FreeColorForeground(x_Disp,x_WinAttr,
		 x_Gc[color_Index]);
		XFreeGC(x_Disp,x_Gc[color_Index]);
	}

	for (tile_Index = 0;tile_Index < XTEEVEE_STATIC_TILE_COUNT;
	 tile_Index++)
	{
		XFreePixmap(x_Disp, tile_Tile[tile_Index]);
	}
}

/* Vertical Roll =========================================================== */
void xteevee_Roll(Display* x_Disp,Window x_Win,XWindowAttributes* x_WinAttr,
      time_t hack_Time,Pixmap hack_Pm)
{
	GC        x_Gc;
	XGCValues x_GcVal;
	int       roll_Y = 0;
	int       roll_Speed = 0;
	int       blank_Height = x_WinAttr->height/10;

	/* Build the GC */
	x_GcVal.foreground = BlackPixel(x_Disp,0);
	x_GcVal.subwindow_mode = IncludeInferiors;
	x_Gc = XCreateGC(x_Disp,x_Win,GCForeground|GCSubwindowMode,&x_GcVal);

	/* Go! */
	if (hack_Time > 0)
	{
		hack_Time += time(NULL);
	}
	while ((roll_Y > 0) || ((time(NULL) < hack_Time) || (hack_Time == 0)))
	{
		if (roll_Y > blank_Height)
		{
			XCopyArea(x_Disp,hack_Pm,x_Win,x_Gc,
			 0,x_WinAttr->height-(roll_Y-blank_Height)-1,
			 x_WinAttr->width,roll_Y-blank_Height,
			 0,0);
		}
		XFillRectangle(x_Disp,x_Win,x_Gc,
		 0,roll_Y-blank_Height,
		 x_WinAttr->width,blank_Height);
		if (roll_Y < x_WinAttr->height)
		{
			XCopyArea(x_Disp,hack_Pm,x_Win,x_Gc,
			 0,0,
			 x_WinAttr->width,x_WinAttr->height-roll_Y,
			 0,roll_Y);
		}

		roll_Y += roll_Speed/2;
		if (roll_Speed < 50)
		{
			roll_Speed++;
		}
		if (roll_Y > x_WinAttr->height+blank_Height)
		{
			roll_Y = 0;
			roll_Speed = 0;
		}

		XSync(x_Disp,0);
                usleep(50000);
		screenhack_handle_events(x_Disp);
	}

	/* Free everything */
	XFreeGC(x_Disp,x_Gc);
}

/* Color-Bars Test Pattern ================================================= */
void xteevee_Bars(Display* x_Disp,Window x_Win,XWindowAttributes* x_WinAttr,
      time_t hack_Time,Pixmap hack_Pm)
{
  draw_colorbars (x_WinAttr->screen, x_WinAttr->visual, x_Win,
                  x_WinAttr->colormap,
                  0, 0, x_WinAttr->width, x_WinAttr->height);

  /* Go! */
  if (hack_Time > 0)
    {
      hack_Time += time(NULL);
    }
  while ((time(NULL) < hack_Time) || (hack_Time == 0))
    {
      screenhack_handle_events(x_Disp);
      usleep(100000);
    }
}


/* Standard XScreenSaver entry point ======================================= */
void screenhack(Display* x_Disp,Window x_Win)
{
	XWindowAttributes x_WinAttr;
        GC                x_Gc;
        XGCValues         x_GcVal;
        Pixmap            screen_Pm;
	time_t            delay_Time;
	int               delay_Mode;
	int               delay_Between;
	int               mode_Index;
	int               mode_Count = 0;
	int               mode_Total = 0;
	char              mode_Arg[64+1];
	int               mode_Min = XTEEVEE_MODE_INCLUDE_IMPLICIT;
	struct
	{
		char* mode_Arg;
		void  (*mode_Func)(Display* x_Disp,Window x_Win,
		      XWindowAttributes* x_WinAttr,time_t hack_Time,
		      Pixmap hack_Pm);
		int   mode_Flag;
	} mode_Mode[] =
	{
		{ XTEEVEE_ARG_STATIC,xteevee_Static,XTEEVEE_MODE_EXCLUDE },
		{ XTEEVEE_ARG_ROLL,  xteevee_Roll,  XTEEVEE_MODE_EXCLUDE },
		{ XTEEVEE_ARG_BARS,  xteevee_Bars,  XTEEVEE_MODE_EXCLUDE },
		{ NULL,              NULL,          -1 }
	};

	/* Grab the screen to give us time to do whatever we want */
	XGetWindowAttributes(x_Disp,x_Win,&x_WinAttr);
        load_random_image (x_WinAttr.screen, x_Win, x_Win, NULL, NULL);

	x_GcVal.subwindow_mode = IncludeInferiors;
	x_Gc = XCreateGC(x_Disp,x_Win,GCSubwindowMode,&x_GcVal);
	screen_Pm = XCreatePixmap(x_Disp,x_Win,x_WinAttr.width,
	 x_WinAttr.height,x_WinAttr.depth);
	XCopyArea(x_Disp,x_Win,screen_Pm,x_Gc,0,0,x_WinAttr.width,
	 x_WinAttr.height,0,0);

	/* Read the arguments */
	delay_Mode = get_integer_resource(XTEEVEE_ARG_DELAY_MODE,"Integer");
	delay_Between = get_integer_resource(XTEEVEE_ARG_DELAY_BETWEEN,
	 "Integer");
	if (!get_boolean_resource(XTEEVEE_ARG_CYCLE,"Boolean"))
	{
		delay_Mode = 0;
	}
	for (mode_Index = 0;mode_Mode[mode_Index].mode_Arg != NULL;
	 mode_Index++)
	{
		if (get_boolean_resource(mode_Mode[mode_Index].mode_Arg,
		 "Boolean") != 0)
		{
			mode_Mode[mode_Index].mode_Flag =
			 XTEEVEE_MODE_INCLUDE_IMPLICIT;
			mode_Count++;
		}
		sprintf(mode_Arg,"_%s",mode_Mode[mode_Index].mode_Arg);
		if (get_boolean_resource(mode_Arg,"Boolean") != 0)
		{
			mode_Mode[mode_Index].mode_Flag =
			 XTEEVEE_MODE_INCLUDE_EXPLICIT;
			mode_Min = XTEEVEE_MODE_INCLUDE_EXPLICIT;
			mode_Count++;
		}
		mode_Total++;
	}
	if (mode_Count == 0)
	{
		fprintf(stderr,"%s: No modes selected!\n",XTEEVEE_NAME);
		return;
	}

	/* Cycle through various modes */
	for (;;)
	{
		/* Find a mode */
		do
		{
			mode_Index = random()%mode_Total;
		} while (mode_Mode[mode_Index].mode_Flag < mode_Min);

		/* Run the hack */
		mode_Mode[mode_Index].mode_Func(x_Disp,x_Win,&x_WinAttr,
		 delay_Mode,screen_Pm);

		/* Restore the screen and wait */
		XCopyArea(x_Disp,screen_Pm,x_Win,x_Gc,0,0,x_WinAttr.width,
		 x_WinAttr.height,0,0);
		delay_Time = time(NULL)+delay_Between;
		while (time(NULL) < delay_Time)
		{
			screenhack_handle_events(x_Disp);
			usleep(100000);
		}
	}
}
